feat: add annotation and label filters to Ambassador Host Source

KyleMartin901 opened the pull request:
This change makes the Ambassador Host source respect the External-DNS annotation-filter and label-filter allowing for an Ambassador Host resource to specify what External-DNS deployment to use when there are multiple External-DNS deployments within the same cluster. Before this change if you had two External-DNS deployments within the cluster and used the Ambassador Host source the first External-DNS to process the resource will create the record and not the one that was specified in the filter annotation.

### Annotation Fillter

I added the `filterByAnnotations` function so that it matched the same way the other sources have implemented annotation filtering. I didn't add the controller check only because I wanted to keep this change to implementing the annotationFilter.

I added Endpoint tests to validate that the filterByAnnotations function works as expected. Again these tests were based of the Endpoint tests that other sources use. To keep the tests simpler I only allow for a single load balancer to be used.

Example: Create two External-DNS deployments 1 public and 1 private and set the Ambassador Host to use the public External-DNS using the annotation filter.

```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: external-dns-private
spec:
  strategy:
    type: Recreate
  selector:
    matchLabels:
      app: external-dns-private
  template:
    metadata:
      labels:
        app: external-dns-private
      annotations:
        iam.amazonaws.com/role: {ARN} # AWS ARN role
    spec:
      serviceAccountName: external-dns
      containers:
      - name: external-dns
        image: k8s.gcr.io/external-dns/external-dns:latest
        args:
        - --source=ambassador-host
        - --domain-filter=example.net # will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
        - --provider=aws
        - --policy=upsert-only # would prevent ExternalDNS from deleting any records, omit to enable full synchronization
        - --aws-zone-type=private # only look at public hosted zones (valid values are public, private or no value for both)
        - --registry=txt
        - --txt-owner-id= {Hosted Zone ID} # Insert Route53 Hosted Zone ID here
        - --annotation-filter=kubernetes.io/ingress.class in (private)
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: external-dns-public
spec:
  strategy:
    type: Recreate
  selector:
    matchLabels:
      app: external-dns-public
  template:
    metadata:
      labels:
        app: external-dns-public
      annotations:
        iam.amazonaws.com/role: {ARN} # AWS ARN role
    spec:
      serviceAccountName: external-dns
      containers:
      - name: external-dns
        image: k8s.gcr.io/external-dns/external-dns:latest
        args:
        - --source=ambassador-host
        - --domain-filter=example.net # will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
        - --provider=aws
        - --policy=upsert-only # would prevent ExternalDNS from deleting any records, omit to enable full synchronization
        - --aws-zone-type= # only look at public hosted zones (valid values are public, private or no value for both)
        - --registry=txt
        - --txt-owner-id= {Hosted Zone ID} # Insert Route53 Hosted Zone ID here
        - --annotation-filter=kubernetes.io/ingress.class in (public)
---
apiVersion: getambassador.io/v3alpha1
  kind: Host
  metadata:
    name: your-hostname
    annotations:
      external-dns.ambassador-service: emissary-ingress/emissary
      kubernetes.io/ingress.class: public
  spec:
		acmeProvider:
      authority: none
		hostname: your-hostname.example.com
```

Fixes kubernetes-sigs/external-dns#2632

### Label Filter

Currently, the `--label-filter` flag can only be used to filter CRDs, Ingress, Service and Openshift Route objects that match the label selector passed through that flag. This change extends the functionality to the Ambassador Host type object.

When the flag is not specified the default value is `labels.Everything()` which is an empty string, the same as before.
Annotation based filter is inefficient because the filtering has to be done in the controller instead of the API server like with label filtering. The Annotation based filtering has been left in for legacy reasons so the Ambassador Host source can be used in conjunction with the other sources that don't yet support label filtering.

It is possible to use label based filtering with annotation based filtering so you can initially filter by label and then filter the returned hosts by annotation. This is not recommended

Fixes kubernetes-sigs/external-dns#2761

**Checklist**

- [X] Unit tests updated
- [ ] End user documentation updated

KyleMartin901 commented:
I have also added the label filter to the Ambassador host source as per suggestion by @alebedev87 in https://github.com/kubernetes-sigs/external-dns/pull/2043#issuecomment-1121664222

alebedev87 commented on source/ambassador_host_test.go:
```diff
@@ -17,26 +17,562 @@ limitations under the License.
+					hostname:  "fake1.org",
+					annotations: map[string]string{
+						"external-dns.ambassador-service": "emissary-ingress/emissary",
+						"kubernetes.io/ingress.class":     "external-ingress",
```
This annotation can be removed as we don't use any annotation filter, just to keep the test to absolute minimum.

KyleMartin901 replied:
Thanks I have removed the `kubernetes.io/ingress.class` annotations from the label [tests](https://github.com/kubernetes-sigs/external-dns/pull/2633/commits/ed61d9bdd84e2f813794bcb1c1c0737e1fc0dfbb) where they weren't being used. I still need to keep the `external-dns.ambassador-service` annotation as that is how External DNS knows to assign the DNS record to the correct Ambassador Host service/endpoint.

alebedev87 commented on source/ambassador_host_test.go:
```diff
@@ -17,26 +17,562 @@ limitations under the License.
+			expected: []*endpoint.Endpoint{},
+		},
+		{
+			title:            "valid matching label filter expression for single host",
```
This test case seems to be a "superset" of `valid matching label filter expression` one, so why not keeping only this one?

KyleMartin901 replied:
I have now removed `valid matching label filter expression` as you are correct already matching valid label within the `valid matching label filter expression for single host` test. 

Sorry I was trying to match tests like what was done within the annotations without really thinking about it.

alebedev87 commented on source/ambassador_host_test.go:
```diff
@@ -17,26 +17,562 @@ limitations under the License.
+		ti := ti
+		t.Run(ti.title, func(t *testing.T) {
+			// Create a slice of Ambassador Hosts
+			ambassadorHosts := make([]*ambassador.Host, 0)
```
Not a big deal at all but I've seen 3 different ways of creating a slice of ambassodor hosts in this PR:
```
var ambassadorHosts []*ambassador.Host
---
filteredList := []*ambassador.Host{}
---
ambassadorHosts := make([]*ambassador.Host, 0)
```
:)

KyleMartin901 replied:
Sorry about that copy and paste issues. Thanks for picking that up for me totally missed it

KyleMartin901 replied:
I have cleaned this up in https://github.com/kubernetes-sigs/external-dns/pull/2633/commits/cfe7728744c41494c02948fc7da18558f02c3cde decided to use the short declaration operator

alebedev87 reviewed (commented):
Overall LGTM, just some cosmetic comments.

KyleMartin901 commented on source/ambassador_host_test.go:
```diff
@@ -17,26 +17,562 @@ limitations under the License.
+		labelSelector labels.Selector
+	}{
+		{
+			title:           "no host",
+			targetNamespace: "",
+			labelSelector: labels.Everything(),
```
Sorry forgot to add the test case in. I was going to test to make sure the Ambassador host was only added if it is within the External DNS targeted namespace like the following sources do
- [Service](https://github.com/kubernetes-sigs/external-dns/blob/master/source/service_test.go#L443)
- [Istio Gateway](https://github.com/kubernetes-sigs/external-dns/blob/master/source/istio_gateway_test.go#L437)
- [Ingress](https://github.com/kubernetes-sigs/external-dns/blob/master/source/ingress_test.go#L449)
- [Contour HTTPProxy](https://github.com/kubernetes-sigs/external-dns/blob/master/source/contour_httpproxy_test.go#L401)

KyleMartin901 replied:
I have added the Target Namespace test now https://github.com/kubernetes-sigs/external-dns/pull/2633/commits/05d9556ad6554439433a886126fab8c61bf26183

KyleMartin901 commented:
@alebedev87 thanks for picking up those issues. 

I wasn't sure on the process of adding in the changes, if you would prefer the changes to be added in as new commits or squashed so the PR still only had the two commits for adding annotation filter and label filter so I have just added them a seperate commits easy to squash if that's what is preferred. Let me know if you would prefer me to squash them or happy as is.

KyleMartin901 commented on source/ambassador_host_test.go:
```diff
@@ -17,26 +17,562 @@ limitations under the License.
+			expectError: true,
+		},
+		{
+			title:            "valid matching annotation filter label",
```
Up to you again copied the test from [Contour](https://github.com/kubernetes-sigs/external-dns/blob/master/source/contour_httpproxy_test.go#L494), [Ingress](https://github.com/kubernetes-sigs/external-dns/blob/master/source/ingress_test.go#L530), [Istio Gateway](https://github.com/kubernetes-sigs/external-dns/blob/master/source/istio_gateway_test.go#L534) and the [service](https://github.com/kubernetes-sigs/external-dns/blob/master/source/service_test.go#L542) tests. Happy to remove it if you would like.

alebedev87 replied:
Since it doesn't bring any added value, I'd prefer to remove it - smaller is simpler for the further maintenance.

KyleMartin901 commented on source/ambassador_host_test.go:
```diff
@@ -17,26 +17,562 @@ limitations under the License.
+		{
+			title:            "valid matching label filter expression",
+			targetNamespace:  "",
+			// annotationFilter: "kubernetes.io/ingress.class in (external-ingress)",
```
Cleaned this up https://github.com/kubernetes-sigs/external-dns/pull/2633/commits/1629661fe18899e067ca5e89b1dfb747f9347119 sorry forgot to go through and make sure was all clean

alebedev87 commented:
@KyleMartin901: I'm not aware of any strict rules about the commits. However I don't think that the commits made to address the review remarks really need to be upstream.

KyleMartin901 commented:
> @KyleMartin901: I'm not aware of any strict rules about the commits. However I don't think that the commits made to address the review remarks really need to be upstream.

I have rebased so there are only 2 commits 1 for adding the annotation filtering and 1 for adding the label filtering with all the changes suggested. Sorry for the delay in getting it all updated.

alebedev87 commented:
/lgtm

alebedev87 commented:
/assign @Raffo

olemarkus commented on source/ambassador_host_test.go:
```diff
@@ -141,3 +658,77 @@ func TestParseAmbLoadBalancerService(t *testing.T) {
 	}
 }
+
+func convertAmbassadorHostToUnstructured(hp *ambassador.Host, s *runtime.Scheme) (*unstructured.Unstructured, error) {
```
I don't feel adding this to the global scope in this file is the right thing to do. There is already logic for converting using this https://github.com/kubernetes-sigs/external-dns/blob/master/source/ambassador_host.go#L121

olemarkus commented on source/ambassador_host_test.go:
```diff
@@ -141,3 +658,77 @@ func TestParseAmbLoadBalancerService(t *testing.T) {
+	return unstructuredAmbassadorHost, nil
+}
+
+func newAmbassadorDynamicKubernetesClient() (*fakeDynamic.FakeDynamicClient, *runtime.Scheme) {
```
If one is not careful and look at the return values, one may be mislead into thinking this creates a non-fake client. Perhaps rename?

olemarkus commented on source/ambassador_host_test.go:
```diff
@@ -141,3 +658,77 @@ func TestParseAmbLoadBalancerService(t *testing.T) {
+	return fakeDynamic.NewSimpleDynamicClient(s), s
+}
+
+type fakeAmbassadorHost struct {
```
This overlaps with this function: https://github.com/kubernetes-sigs/external-dns/blob/master/source/ambassador_host_test.go#L86
Suggest refactor to use one or the other.

olemarkus commented on source/ambassador_host_test.go:
```diff
@@ -141,3 +658,77 @@ func TestParseAmbLoadBalancerService(t *testing.T) {
+	return host
+}
+
+type fakeAmbassadorLoadBalancerService struct {
```
I don't see anything ambassador-specific in this struct. Can't one reuse https://github.com/kubernetes-sigs/external-dns/blob/0bf5faed42215ce20533de385bb6dccde10c4ab1/source/contour_httpproxy_test.go#L52 ?

Possibly move this to e.g https://github.com/kubernetes-sigs/external-dns/blob/0bf5faed42215ce20533de385bb6dccde10c4ab1/source/source_test.go

KyleMartin901 commented:
/remove-lifecycle stale

KyleMartin901 commented:
/remove-lifecycle stale

mloiseleur commented:
@KyleMartin901 I understand it has been a long time but do you think you can update the [sources part of doc](https://kubernetes-sigs.github.io/external-dns/v0.14.0/sources/sources/) and rebase your PR ? It's looking good.

mloiseleur commented:
Since author is not answering, I'm closing this PR.
Feel free to re-open it or open a new one if you need it.
/close

KyleMartin901 commented:
Sorry I am just getting back from 6 weeks holiday where I had my notifications turned off. I’ll fix up the issues this weekend  @mloiseleur

mloiseleur commented:
sure, no problem
/reopen

KyleMartin901 commented:
@mloiseleur thanks for reopening the PR and giving me the time to rebase and update the docs. I have now pushed those changes, let me know if there is anything else you would like me to do.

mloiseleur commented:
/ok-to-test
/retitle feat: add annotation and label filters to Ambassador Host Source

KyleMartin901 commented:
@mloiseleur sorry about the failed tests, I have fixed them up by adding the now required record type. Sorry I was lazy and just did the rebase and didn't rerun the tests but I have now run the tests and they are all passing. I added a lazy commit let me know if you want it squashed in or changed.

mloiseleur commented:
/lgtm
/label tide/merge-method-squash

mloiseleur commented:
@KyleMartin901 This PR has been approved. It means that it just needs a rebase, to pass tests and it will be merged.

KyleMartin901 commented:
Thanks @mloiseleur i will attempt to get this done this week. Looks like someone got tests merged in before mine so going to refactor to match what has already been merged

mloiseleur commented:
/remove-lifecycle rotten
